lib/UserBackend.php

```diff
@@ -677,9 +710,51 @@ public function updateAttributes($uid,
 					$groupManager->get($group)->removeUser($user);
 				}
 			}
+
+			if ($newAvatar !== null) {
+                $image = new \OCP\Image();
+                $fileData = file_get_contents($newAvatar);
+                $image->loadFromData($fileData);
+                $this->setOwnCloudAvatar($uid, $image);
+            }
 		}
 	}
 
+    private function setOwnCloudAvatar($uid, $image) {
+        if (!$image->valid()) {
+            $this->logger->debug('avatar image data from LDAP invalid for ' . $uid);
+            return false;
+        }
+
+
+        //make sure it is a square and not bigger than 128x128
+        $size = min([$image->width(), $image->height(), 128]);
+        if (!$image->centerCrop($size)) {
+            $this->logger->debug('croping image for avatar failed for ' . $uid);
+            return false;
+        }
+
+        $fs = new FilesystemHelper();
+        if (!$fs->isLoaded()) {
+            $fs->setup($uid);
+        }
+
+        try {
+            $avatarManager = \OC::$server->getAvatarManager();
+
+            $avatar = $avatarManager->getAvatar($uid);
+            $avatar->set($image);
```

Comment:
this writes the avatar every time. Especially when auth is happening often, this is a waste of resources and expensive (since it triggers several updates in other components).

Reply:
That is right. The avatar is updated every time any other SAML attribute is updated. I think this behaviour should be the same as in the LDAP extension. I agree that it might create unnecessary overhead. Do you have a suggestion how to improve this?

Reply:
In LDAP we store the and compare the hash of the image data. Only if it does not match, the received image data is set as new avatar.

it might be useful to move this into the avatar api to avoid duplication, but now this would earliest go into NC 23.

Reply:
can we use duplicate code here with an annotation to refactor it for NC23 (if the code moves into the avatar api)?

Reply:
I changed the code to save the checksum of the avatar image in user data and compare it before any image processing is done, when the user logs in.